src/shapes/src/SmoothCorner.cpp

```diff
@@ -0,0 +1,138 @@
+    }
+
+    // point 1 and 2 are the end points of the shape where the cylinder
+    // intersects the cuboid. They also define normals of the cuboid side planes
+    auto to_point_1 =
+        m_radius * vec_rotate(m_axis, -0.5 * m_delta_phi, m_direction);
+    auto to_point_2 =
+        m_radius * vec_rotate(m_axis, 0.5 * m_delta_phi, m_direction);
+
+    if (pos_delta_phi <= -0.5 * m_delta_phi) {
+      vec = pos_projected - to_point_1;
+      dist = vec.norm();
+    } else {
+      if (pos_delta_phi >= 0.5 * m_delta_phi) {
+        vec = pos_projected - to_point_2;
+        dist = vec.norm();
+      } else {
+        if (pos_norm < m_radius) {
+          dist = m_radius - pos_norm;
+          vec = -dist * pos_projected / pos_norm;
+        } else {
+          // calc distance to the planes and the cylinder, then find the minimum
+          auto dist_plane_1 = (pos_projected * to_point_1 / to_point_1.norm()) -
+                              to_point_1.norm();
+          auto vec_plane_1 = dist_plane_1 * to_point_1 / to_point_1.norm();
+          auto dist_plane_2 = (pos_projected * to_point_2 / to_point_2.norm()) -
+                              to_point_2.norm();
+          auto vec_plane_2 = dist_plane_2 * to_point_2 / to_point_2.norm();
+
+          // find closest plane segment or take the intersection of the segments
+          double dist_plane;
+          Utils::Vector3d vec_plane;
+          if (dist_plane_1 >= 0. and dist_plane_2 >= 0.) {
+            vec_plane = pos_projected - std::sqrt(2.) * m_radius * m_direction;
+            dist_plane = vec_plane.norm();
+          } else {
+            if (dist_plane_1 < 0. and dist_plane_2 < 0.) {
+              bool plane_1_is_closer =
+                  Utils::abs(dist_plane_1) < Utils::abs(dist_plane_2);
+              dist_plane = plane_1_is_closer ? dist_plane_1 : dist_plane_2;
+              vec_plane = plane_1_is_closer ? vec_plane_1 : vec_plane_2;
+            } else {
+              bool plane_1_is_positive = dist_plane_1 >= 0;
+              dist_plane = plane_1_is_positive ? dist_plane_1 : dist_plane_2;
+              vec_plane = plane_1_is_positive ? vec_plane_1 : vec_plane_2;
+            }
+          }
+
+          auto dist_cyl = m_radius - pos_norm;
+          auto vec_cyl = -dist_cyl * pos_projected / pos_norm;
+
+          if (Utils::abs(dist_cyl) < Utils::abs(dist_plane)) {
+            dist = dist_cyl;
+            vec = vec_cyl;
+          } else {
+            dist = dist_plane;
+            vec = vec_plane;
+          }
+        }
+      }
+    }
+  } // 2D end, now to the 3rd dimension
+
+  auto upper_lower = rel_z > 0 ? 1 : -1;
+  auto dist_top_bottom = Utils::abs(rel_z) - 0.5 * m_height;
+
+  // if inside the shape, check if the bottom or top wall are closer
+  if (dist < 0. and dist_top_bottom < 0) {
+    if (Utils::abs(dist_top_bottom) < Utils::abs(dist)) {
+      dist = dist_top_bottom;
+      vec = upper_lower * dist * m_axis;
+    }
+  }
```

Comment:
This expression can be evaluated by two comparisons
```suggestion
  if (dist_top_bottom < 0 and dist_top_bottom > dist) {
      dist = dist_top_bottom;
      vec = upper_lower * dist * m_axis;
  }
```